### PR TITLE
Optimizing ml models with optuna

### LINKED ISSUE #5749: Optimizing ML Models with Optuna
### Topic Suggestion
Optimizing ML Models with Optuna
#### Proposed article introduction
Optuna is the SOTA algorithm used to make changes to ML and deep learning models. It all depends on the Bayesian fine-tuning method. It stops trying combinations that don't improve our score any more and only tries combinations that do.

People who work on ML models have a hard time. In order to make it easier for us to fine-tune our model, there are a lot of tools out there. You need to make sure that our data is important and that our ML model fits on it perfectly. Here, we'll learn how to make an Xgboost regressor work even better.

When you use machine learning, a hyperparameter is a parameter that can be changed to change how the learning process works.  The values of hyperparameters can't be guessed from the data. It's important to set these parameters so that the model can learn from our data. You can change the learning parameters so that the model fits our data and learns from it to make smart predictions.
#### Key takeaways 
- Visualization of Tuning
- Comparing the Tuned Model and Base Model
- Creating Hyper-Parameters Space

#### Article quality
In this article, we'll learn how to improve ML models and how to find the best parameters for the models we make. The code will be easy to read and understand.
#### References
N/A
### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by NdoloBonface**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

`This closes #5749 

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by NdoloBonface** [comment]
@srishilesh 
Duplicate of #6816

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
+Software framework Optuna is used to automate optimization processes. It uses a variety of samplers, including grid search, random, bayesian, and evolutionary algorithms, to find the best values for hyperparameters. If you use machine learning methods, a `hyperparameter` is a parameter that may be changed to influence the learning process. Using our data, we'll tweak the model's learning parameters.
```
[the remark this answers is not in the document]
Comment (srishilesh) [reply content/articles/optimizing-ml-models-with-optuna/index.md:1]: `samplers`?

[the remark this answers is not in the document]
Comment (srishilesh) [reply content/articles/optimizing-ml-models-with-optuna/index.md:1]: Keep the introduction simple. Avoid these terminolgies.

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+It is possible to use various resources to improve our model. Ensure that the ML model is acceptable for our data and required. The following section will discuss enhancements and new applications for `Xgboost` regressors which is an open-source library that implements the gradient boosting algorithm in an effective manner.
+### Prerequisites
+- Have a basic knowledge of [Flutter](https://flutter.dev/).
+- Have flutter installed in your IDE.
+### Table of contents
+- [Reasons for using Optuna](#reasons-for-using-optuna)
+- [Optuna functionalities](#optuna-functionalities)
+- [Optimization process](#optimization-process)
+- [Conclusion](#conclusion)
+### Reasons for using Optuna
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:12]: What are you trying to optimize? State the problem before jumping into solution

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+- The length of the optimization procedure may be specified.
+- Pandas DataFrame integration.
+- Pruning is a technique that the algorithm uses to eliminate poor-quality samples quickly.
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:15]: How is pruning related to optuna?

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+The Optuna software framework automates the hyperparameter optimization process. **Tree-structured Parzen Estimator** is the name of the `Bayesian` optimization algorithm it uses. According to Bayesian fine-tuning, once it stops improving our score anymore, it stops looking for new ones.
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:17]: `Tree-structured Parzen Estimator`

Add reference link

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+### Optuna functionalities
+- Installation is quick and easy, and there are only a few prerequisites to meet.
+- To conduct large-scale investigations, optuna provides minimal or no code modifications.
+- Analyzes optimization results using a wide range of graphing tools.
+### Optimization process
+#### Importing necessary modules
+We will build a base model by first incorporating essential modules and data into a dataframe. 
+
+1. **Loading the dataset**
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:27]: Use `#####` instead of numbers

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+ A pandas data frame will be used to segregate features and targets when the data is returned as a matrices by `load boston()`.
+```python
+Our_loadboston = load_boston()
+dataframe = pan.DataFrame(Our_loadboston.data , columns = Our_loadboston.feature_names)
+dataframe['target'] = Our_loadboston.target
+X = dataframe.iloc[:,dataframe.columns != 'target']
+y = dataframe.target
+from sklearn.preprocessing import StandardScaler
+standardSc = StandardScaler()
+X = standardSc.fit_transform(X)
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:47]: You have not explaind the code as I had requested previously

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+```
+3. **Dataset splitting**
+
+By dividing the dataset into train and test, the model is better able to reach a convergent state.
+```python
+from sklearn.model_selection import train_test_split
+X_train, X_test, y_train,y_test = train_test_split(X, y, test_size = 0.2, random_state = 12)
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:54]: You have not explaind the code as I had requested previously

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+```python
+import xgboost as gbst
+from sklearn.model_selection import cross_val_score
+xgboo_reg = gbst.XGBRegressor()
+# Cross-validation is a method for evaluating a score
+ourScores = cross_val_score(xgboo_reg, X_train,y_train , scoring = 'neg_root_mean_squared_error',
+                            #using all available CPUs
+                            n_jobs = -1,cv = 10)
+print(num.mean(ourScores), num.std(ourScores))
+print(ourScores)
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:70]: You have not explaind the code as I had requested previously

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+```
+3. **Dataset splitting**
+
+By dividing the dataset into train and test, the model is better able to reach a convergent state.
+```python
+from sklearn.model_selection import train_test_split
+X_train, X_test, y_train,y_test = train_test_split(X, y, test_size = 0.2, random_state = 12)
+```
+#### Testing the model
+We can see how much better our model is after fine-tuning the parameters by testing the model. We will use the `cross_val_score()` method to evaluate a score and xboost for gradient boosting algorithm.
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:57]: Which model are you testing?

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+                'min_child_weight':trial.suggest_int('min_child_weight',0,5),
+                'learning_rate':trial.suggest_loguniform('learning_rate',0.05,0.5),
+               #log distribution between [0.05,.5]
+                'colsample_bytree':trial.suggest_uniform('colsample_bytree',0.4,0.9),
+                'max_depth':trial.suggest_int('max_depth', 3, 5),
+                'nthread' : -1
+            }
+  #returns the regressionModelse score
+  return(return_score(parameter))
+```
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:121]: Explain

**Review by srishilesh — CHANGES_REQUESTED** [review]
Hey @NdoloBonface ,

You have not addressed my suggestions. Please go through them carefully.

I'm sure you wouldn't have gone through other published articles to see how other authors have done it.

Fyi, this is your 1st duplicate PR, if all the requested changes are not done in the new PR, we might have to close this topic. By closing the topic, you might have to suggest a new one.

Ensure that your next PR must be revisions free.

As a few guideline:
- Explain all the terminologies. Add reference links for it
- State the "why" and "how" of everything you do.
- Explain code snippets
- Don't just put your thoughts into words. Instead, think from the reader's perspective and write them.

Take help from other contributors if needed to fine tune your article.

Ping me in the new PR.

Thanks.

**Comment by NdoloBonface** [comment]
@srishilesh ill work on it and ping you. For now I am held up with exams.

**Comment by srishilesh** [comment]
> @srishilesh ill work on it and ping you. For now I am held up with exams.
> 

@NdoloBonface 
Sure. No problem. Take your time.

Let me know if you need any help.

**Comment by NdoloBonface** [comment]
> Reference in new issue
> 
> ### Reference in new issue
> Repository
> engineering-education 
> Title

I will ping you via slack
Thanks
